### DISCUSSION TITLE
Custom merge that doesn't override values with undefined

### DISCUSSION THREAD

**Discussion by RebeccaStevens** [RebeccaStevens/deepmerge-ts#25] [Show and tell]
```ts
import {
  deepmergeCustom,
  type DeepMergeLeaf,
  type DeepMergeMergeFunctionsURIs,
} from "deepmerge-ts";

export const deepmergeDefined = deepmergeCustom<{
  DeepMergeOthersURI: "DeepMergeLeafNoUndefinedOverrideURI";
}>({
  mergeOthers: (values, utils) =>
    utils.defaultMergeFunctions.mergeOthers(
      values.filter((v) => v !== undefined)
    ),
});


declare module "deepmerge-ts" {
  interface DeepMergeMergeFunctionURItoKind<
    Ts extends ReadonlyArray<unknown>,
    MF extends DeepMergeMergeFunctionsURIs,
    in out M
  > {
    readonly DeepMergeLeafNoUndefinedOverrideURI: DeepMergeLeafNoUndefinedOverride<Ts>;
  }
}

type DeepMergeLeafNoUndefinedOverride<Ts extends ReadonlyArray<unknown>> =
  DeepMergeLeaf<FilterOutUnderfined<Ts>>;

type FilterOutUnderfined<T extends ReadonlyArray<unknown>> =
  FilterOutUnderfinedHelper<T, []>;

type FilterOutUnderfinedHelper<
  T extends ReadonlyArray<unknown>,
  Acc extends ReadonlyArray<unknown>
> = T extends readonly []
  ? Acc
  : T extends readonly [infer Head, ...infer Rest]
  ? Head extends undefined
    ? FilterOutUnderfinedHelper<Rest, Acc>
    : FilterOutUnderfinedHelper<Rest, [...Acc, Head]>
  : T;
```

```ts
import { deepmergeDefined} from "./path/to/file";

const alex = {
  name: {
    first: "Alex",
    last: "Alexson",
  },
  pets: ["Cat", "Parrot"],
};

const tony = {
  name: {
    first: undefined,
    last: "Tonison",
  },
  pets: ["Dog"],
};

const result = deepmergeDefined(alex, tony);
console.log(result);

// {
//   name: {
//     first: "Alex"
//     last: "Tonison"
//   }
//   pets: [
//     "Cat"
//     "Parrot"
//     "Dog"
//   ]
// }
```

**Comment by percy507** [comment]
> ```ts
> export const myDeepmerge = deepmergeCustom<{
>   DeepMergeOthersURI: "DeepMergeLeafNoUndefinedOverrideURI";
> }>({
>   mergeOthers: (values, utils) =>
>     utils.defaultMergeFunctions.mergeOthers(
>       values.filter((v) => v !== undefined),
>       utils
>     )
> });
> ```

@RebeccaStevens The example seems have error.

<img width="1002" alt="image" src="https://github.com/RebeccaStevens/deepmerge-ts/assets/18365078/1e0cd429-4f41-4532-8c1c-ac0e9e8a1261">

**Reply by RebeccaStevens** [reply]
I've just updating the code above to now work with the latest version of the library. Try the new example.

**Reply by percy507** [reply]
Works fine. Thank u.

But I still have a problem.  I add `mergeArrays: false,` option to the example and make a test. I find the merged type is not right and confused to how to fix it. 

See below image, the type of `a` expected be `string[]`.

<img width="624" alt="image" src="https://github.com/RebeccaStevens/deepmerge-ts/assets/18365078/c16a1956-86bc-4dae-9a3e-6b2fa9353b3a">

**Reply by percy507** [reply]
Ah, find it. Thanks again.

<img width="848" alt="image" src="https://github.com/RebeccaStevens/deepmerge-ts/assets/18365078/18d0a85f-e5fb-49a1-b525-c21e6907efe1">

**Comment by percy507** [comment]
Continue the above example with the additional option `mergeArrays: false`.  

I create an usual example: merge `default options` and `input options` in a class constructor. But the merged type is unexpected. Then I have to use type assertion. I wonder how to fix this.

```ts
import { deepmerge } from './deepmerge';

interface Options {
  opt1?: string;
  opt2?: boolean;
  opt3?: boolean;
  opt4?: string[];
}

const defaultOptions = {
  opt3: true,
  opt4: [] as string[],
};

type MergedOptions = typeof defaultOptions & Options;

export class MyEditor {
  #options;
  #options2;

  constructor(options: Options) {
    this.#options = deepmerge(defaultOptions, options);

    // ==> Wrong: vscode recognizes `opt4` as `string[] | undefined`
    if (this.#options.opt4.length > 2) {
      // do something
    }

    this.#options2 = deepmerge(defaultOptions, options) as MergedOptions;

    // ==> Expected: vscode recognizes `opt4` as `string[]`
    if (this.#options2.opt4.length > 2) {
      // do something
    }
  }
}
```

**Reply by RebeccaStevens** [reply]
Looks like you have found a bug/limitation of types of this custom deepmerge. I'll need to spend some time investigating to come up with a fix. For now, you it's probably best to just use `as MergedOptions` as you are.

An alternative workaround would be to un-optional-ize the `Options` type with something like:

```ts
type Unoptionalize<T> = {
  [K in OptionalKeys<T>]-?: T[K] | undefined;
} & {
  [K in RequiredKeys<T>]: T[K];
};

type KeysOfType<T, U> = {
  [K in keyof T]: T[K] extends U ? K : never;
}[keyof T];

type RequiredKeys<T> = Exclude<
  KeysOfType<T, Exclude<T[keyof T], undefined>>,
  undefined
>;

type OptionalKeys<T> = Exclude<keyof T, RequiredKeys<T>>;
```

```ts
this.#options = deepmerge(defaultOptions, options as Unoptionalize<Options>);
```

I'll probably need to somehow incorporate these types into the actual fix.
